### PR TITLE
feat(dgw): stabilize /jet/jrec/shadow route

### PR DESCRIPTION
**Pull request by irvingoujAtDevolution**

### CODE REVIEW

File: devolutions-gateway/src/api/jrec.rs [hunk 1]
```diff
@@ -500,25 +497,17 @@ async fn shadow_recording(
...
-    let path = conf_handle
-        .get_conf()
-        .recording_path
-        .join(id.to_string())
-        .join(filename);
+    let recording_dir_path = conf_handle.get_conf().recording_path.join(id.to_string());
 
-    if !path.exists() || !path.is_file() {
+    if !recording_dir_path.exists() || !recording_dir_path.is_dir() {
         return Err(HttpError::not_found().msg("requested file does not exist"));
```
Comment (CBenoit) [inline devolutions-gateway/src/api/jrec.rs:511]: nitpick: Update the error message

File: devolutions-gateway/src/streaming.rs [hunk 2]
```diff
@@ -17,22 +21,23 @@ impl streamer::Signal for ShutdownSignal {
...
+    recording_folder_path: camino::Utf8PathBuf,
     ws: axum::extract::WebSocketUpgrade,
     shutdown_notify: Arc<Notify>,
     recordings: crate::recording::RecordingMessageSender,
     recording_id: Uuid,
 ) -> anyhow::Result<Response<Body>> {
+    let recording_file_path = get_recording_file_path(recording_folder_path)?;
     // 1.identify the file type
-    if path.extension() != Some(RecordingFileType::WebM.extension()) {
+    if recording_file_path.extension() != Some(RecordingFileType::WebM.extension()) {
```
Comment (CBenoit) [inline devolutions-gateway/src/streaming.rs:32]: issue: Are you sure about checking the extension now that this is a folder…?

File: devolutions-gateway/src/streaming.rs [hunk 3]
```diff
@@ -73,5 +78,13 @@ pub(crate) async fn stream_file(
...
-    Ok(upgrade_result)
+    return Ok(upgrade_result);
+
+    fn get_recording_file_path(recording_folder_path: camino::Utf8PathBuf) -> anyhow::Result<camino::Utf8PathBuf> {
+        let manifest = JrecManifest::read_from_file(&recording_folder_path.join("recording.json"))?;
+        // TODO: handle multiple files
+        let file = manifest.files.get(0).context("no files in manifest")?;
+
+        Ok(recording_folder_path.join(&file.file_name))
+    }
```
Comment (CBenoit) [inline devolutions-gateway/src/streaming.rs:89]: issue: Maybe we want the last (current) file? You can use `.last()` for that. Also rename the function to `get_last_recording_file` to reflect that.

File: devolutions-gateway/src/recording.rs [hunk 4]
```diff
@@ -29,23 +29,23 @@ const BUFFER_WRITER_SIZE: usize = 64 * 1024;
 
 #[derive(Debug, Clone, Serialize, Deserialize)]
 #[serde(rename_all = "camelCase")]
-struct JrecFile {
-    file_name: String,
+pub(crate) struct JrecFile {
+    pub(crate) file_name: String,
```
Comment (CBenoit) [inline devolutions-gateway/src/recording.rs:33]: issue: Let’s not leak the private API.

**Review by CBenoit — COMMENTED** [review]
The up to date manifest is stored part of the `OnGoingRecording` struct:

https://github.com/Devolutions/devolutions-gateway/blob/c91f46941918ed89f419b8d04f66bbaf3611be18/devolutions-gateway/src/recording.rs#L206

I suggest you add a new message to retrieve the path to the current recording file instead of duplicating the logic.

File: devolutions-gateway/src/streaming.rs [hunk 5]
```diff
@@ -17,22 +21,23 @@ impl streamer::Signal for ShutdownSignal {
...
+    if recording_file_path.extension() != Some(RecordingFileType::WebM.extension()) {
         anyhow::bail!("invalid file type");
     }
     // 2.if the file is actively being recorded, then proceed
     let Ok(Some(OnGoingRecordingState::Connected)) = recordings.get_state(recording_id).await else {
         anyhow::bail!("file is not being recorded");
     };
 
-    let streaming_file = ReOpenableFile::open(&path).with_context(|| format!("failed to open file: {path:?}"))?;
+    let streaming_file = ReOpenableFile::open(&recording_file_path)?;
```
[the remark this answers is not in the document]
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/streaming.rs:40]: it's redundant,tokio::fs will give enough information

File: devolutions-gateway/src/streaming.rs (hunk 3 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/streaming.rs]: updated

File: devolutions-gateway/src/recording.rs (hunk 4 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/recording.rs]: updated

**Review by CBenoit — APPROVED** [review]

**Pull request merged**